compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUMaterializeEncodingPass.cpp

```diff
@@ -175,10 +270,35 @@ void LLVMCPUMaterializeEncodingPass::runOnOperation() {
   }
 }
 
+void LLVMCPUMaterializeUpperBoundTileSizePass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  auto operation = getOperation();
+  auto targetAttrs =
+      IREE::HAL::DeviceTargetAttr::lookupExecutableTargets(operation);
+  RewritePatternSet patterns(context);
+  MaterializeEncodingFn materializeEncodingFn =
+      getUpperBoundMaterializeEncodingFn(targetAttrs);
+  if (!materializeEncodingFn) {
+    return signalPassFailure();
+  }
+  populateMaterializeUpperBoundTileSizePatterns(patterns,
```

Comment:
I think you made this a separate pass because you want to iterate over a array of targets and pick a pad size? I'd assume that you would do the same for the actual inner tile dimensions of the packing operation. So this seems a bit premature. Maybe we can drop this extra pass and just incorporate this into the existing pass?

Reply:
The primary reason why this needs to be a separate pass is that MaterializeEncoding runs on HAL executables (dispatch functions, which are where `set_encoding` ops are found at that point), while this needs to run on "top-level" code that is outside of HAL executables (the function that makes the flow.dispatch calls).

While these `upper_bound_tile_size` ops are initially found on both sides (both inside and outside of HAL executables), the ones that are in HAL executables are just dropped as the `pad + set_encoding` pairs of ops get rewritten into just `tensor.pack`, the `pad` op goes unused, and it was the only thing there that was using the `upper_bound_tile_size`).  By contrast, the `upper_bound_tile_size` on the outside of HAL executables are consumed by flow.dispatch ops, so they don't go away, so that's where we need this pass to run on them.

It's still possible that one could find a way to unify these two passes, just saying why it's not completely trivial, it would bring more refactoring into the scope of this PR.